File: android/app/src/main/java/app/trustcam/Pdq.kt
package app.trustcam

import kotlin.math.cos
import kotlin.math.sqrt

/**
 * PDQ perceptual hash — line-by-line port of spikes/pdq_ref.py (itself a port
 * of the official ThreatExchange C++): two Jarosz box-filter passes, 64x64
 * center-sample decimation, 16x64 DCT with the (i+1) row offset, Torben
 * median, bit k = DCT coeff (k/16, k%16) > median. Float (32-bit) arithmetic
 * throughout to keep bit parity with the python reference; the quality metric
 * is intentionally omitted (unused on device).
 */
object Pdq {
    private const val OUT = 64 // downsample edge
    private const val DCT_N = 16 // kept DCT coefficients per axis

    /** First 104 bits of the 256-bit hash, packed MSB-first into 13 bytes. */
    fun hash104(luma: FloatArray, width: Int, height: Int): ByteArray {
        val bits = hash256(luma, width, height)
        val out = ByteArray(13)
        for (k in 0 until 104) {
            if (bits[k] == 1) {
                out[k ushr 3] = (out[k ushr 3].toInt() or (1 shl (7 - (k and 7)))).toByte()
            }
        }
        return out
    }

    /** Full 256-bit hash as 0/1 ints; luma is row-major, values 0..255. */
    internal fun hash256(luma: FloatArray, width: Int, height: Int): IntArray {
        require(luma.size == width * height) { "luma size does not match dimensions" }
        val b64 = if (width == OUT && height == OUT) {
            luma.copyOf()
        } else {
            val buf = luma.copyOf()
            // computeJaroszFilterWindowSize(dim, 64) = ceil(dim / 128)
            jarosz(buf, height, width, (width + 127) / 128, (height + 127) / 128)
            decimate(buf, height, width)
        }
        val dct = dct16(b64)
        val median = torben(dct)
        return IntArray(DCT_N * DCT_N) { if (dct[it] > median) 1 else 0 }
    }

    /**
     * Sliding unnormalized box filter — same four phases as the reference
     * (grow window, first full outputs, slide, shrink window).
     */
    private fun box1d(
        inv: FloatArray, inOff: Int, inStride: Int,
        out: FloatArray, outOff: Int, outStride: Int,
        n: Int, fullWindow: Int
    ) {
        val half = (fullWindow + 2) / 2
        var li = 0
        var ri = 0
        var oi = 0
        var s = 0f
        var win = 0
        repeat(half - 1) {
            s += inv[inOff + ri * inStride]
            win++
            ri++
        }
        repeat(fullWindow - half + 1) {
            s += inv[inOff + ri * inStride]
            win++
            out[outOff + oi * outStride] = s / win
            ri++
            oi++
        }
        repeat(n - fullWindow) {
            s += inv[inOff + ri * inStride]
            s -= inv[inOff + li * inStride]
            out[outOff + oi * outStride] = s / win
            li++
            ri++
            oi++
        }
        repeat(half - 1) {
            s -= inv[inOff + li * inStride]
            win--
            out[outOff + oi * outStride] = s / win
            li++
            oi++
        }
    }

    /** Two passes of row + column box filters (Jarosz approximation of Gaussian). */
    private fun jarosz(buf: FloatArray, rows: Int, cols: Int, wRow: Int, wCol: Int) {
        val tmp = FloatArray(buf.size)
        repeat(2) {
            for (i in 0 until rows) box1d(buf, i * cols, 1, tmp, i * cols, 1, cols, wRow)
            for (j in 0 until cols) box1d(tmp, j, cols, buf, j, cols, rows, wCol)
        }
    }

    /** Center sampling down to 64x64 — same index formula as the reference. */
    private fun decimate(buf: FloatArray, rows: Int, cols: Int): FloatArray {
        val out = FloatArray(OUT * OUT)
        for (oi in 0 until OUT) {
            val ii = ((oi + 0.5) * rows / OUT).toInt()
            for (oj in 0 until OUT) {
                val jj = ((oj + 0.5) * cols / OUT).toInt()
                out[oi * OUT + oj] = buf[ii * cols + jj]
            }
        }
        return out
    }

    // 16x64 DCT-II basis; PDQ keeps coefficients 1..16 hence the (i+1) factor.
    private val dctMat: FloatArray by lazy {
        val scale = sqrt(2.0 / OUT).toFloat()
        FloatArray(DCT_N * OUT).also { m ->
            for (i in 0 until DCT_N) {
                for (j in 0 until OUT) {
                    m[i * OUT + j] = scale *
                        cos(Math.PI / 2.0 / OUT * (i + 1) * (2 * j + 1)).toFloat()
                }
            }
        }
    }

    /**
     * fma(a, b, s) in float32, emulated via double (exact: a*b fits a double).
     * The reference BLAS accumulates dot products with sequential fused
     * multiply-adds; plain float32 multiply-then-add flips near-median bits
     * on smooth images, so we must round the same way.
     */
    private fun fma(a: Float, b: Float, s: Float): Float =
        (a.toDouble() * b.toDouble() + s.toDouble()).toFloat()

    /** B = D x b64 x D^T (16x16), float32 with fused accumulation. */
    private fun dct16(b64: FloatArray): FloatArray {
        val d = dctMat
        val t = FloatArray(DCT_N * OUT)
        for (i in 0 until DCT_N) {
            for (j in 0 until OUT) {
                var s = 0f
                for (k in 0 until OUT) s = fma(d[i * OUT + k], b64[k * OUT + j], s)
                t[i * OUT + j] = s
            }
        }
        val out = FloatArray(DCT_N * DCT_N)
        for (i in 0 until DCT_N) {
            for (j in 0 until DCT_N) {
                var s = 0f
                for (k in 0 until OUT) s = fma(t[i * OUT + k], d[j * OUT + k], s)
                out[i * DCT_N + j] = s
            }
        }
        return out
    }

    /** Torben median — no reordering, exact same guess/bound updates as the C++. */
    private fun torben(m: FloatArray): Float {
        val n = m.size
        val halfN = (n + 1) / 2
        var lo = m[0]
        var hi = m[0]
        for (v in m) {
            if (v < lo) lo = v
            if (v > hi) hi = v
        }
        while (true) {
            val guess = (lo + hi) / 2
            var less = 0
            var greater = 0
            var maxlt = lo
            var mingt = hi
            for (v in m) {
                if (v < guess) {
                    less++
                    if (v > maxlt) maxlt = v
                } else if (v > guess) {
                    greater++
                    if (v < mingt) mingt = v
                }
            }
            if (less <= halfN && greater <= halfN) {
                val equal = n - less - greater
                return when {
                    less >= halfN -> maxlt
                    less + equal >= halfN -> guess
                    else -> mingt
                }
            }
            if (less > greater) hi = maxlt else lo = mingt
        }
    }
}

File: web/js/pdq.js
// PDQ perceptual hash — JS port of spikes/pdq_ref.py (itself a line-by-line
// port of the official ThreatExchange pdq/cpp). Math.fround after every
// arithmetic step mirrors the reference's float32 pipeline, so bits match
// the Python vectors within 0-1 flips near the median.
;(typeof window !== 'undefined' ? window : globalThis).TrustCamPdq = (() => {
  const REC601 = [0.299, 0.587, 0.114]
  const f32 = Math.fround

  // one pass of the Jarosz box filter along a vector (four phases: ramp-up,
  // initial full windows, steady state, ramp-down)
  function box1d (invec, inOff, inStride, outvec, outOff, outStride, n, fullWindow) {
    const half = Math.floor((fullWindow + 2) / 2)
    const p1 = half - 1
    const p2 = fullWindow - half + 1
    const p3 = n - fullWindow
    const p4 = half - 1
    let li = 0
    let ri = 0
    let oi = 0
    let s = 0
    let win = 0
    for (let k = 0; k < p1; k++) {
      s = f32(s + invec[inOff + ri * inStride])
      win++
      ri++
    }
    for (let k = 0; k < p2; k++) {
      s = f32(s + invec[inOff + ri * inStride])
      win++
      outvec[outOff + oi * outStride] = f32(s / win)
      ri++
      oi++
    }
    for (let k = 0; k < p3; k++) {
      s = f32(s + invec[inOff + ri * inStride])
      s = f32(s - invec[inOff + li * inStride])
      outvec[outOff + oi * outStride] = f32(s / win)
      li++
      ri++
      oi++
    }
    for (let k = 0; k < p4; k++) {
      s = f32(s - invec[inOff + li * inStride])
      win--
      outvec[outOff + oi * outStride] = f32(s / win)
      li++
      oi++
    }
  }

  // two reps of row-then-column box filtering, in place on buf (rows*cols)
  function jarosz (buf, rows, cols, wrow, wcol) {
    const tmp = new Float32Array(rows * cols)
    for (let rep = 0; rep < 2; rep++) {
      for (let i = 0; i < rows; i++) {
        box1d(buf, i * cols, 1, tmp, i * cols, 1, cols, wrow)
      }
      for (let j = 0; j < cols; j++) {
        box1d(tmp, j, cols, buf, j, cols, rows, wcol)
      }
    }
  }

  // center-sample decimation to 64x64
  function decimate (buf, rows, cols) {
    const out = new Float32Array(64 * 64)
    for (let oi = 0; oi < 64; oi++) {
      const ii = Math.floor(((oi + 0.5) * rows) / 64)
      for (let oj = 0; oj < 64; oj++) {
        const jj = Math.floor(((oj + 0.5) * cols) / 64)
        out[oi * 64 + oj] = buf[ii * cols + jj]
      }
    }
    return out
  }

  // 16x64 DCT-II matrix with the PDQ (i+1) row offset, float32 entries
  const DCT = (() => {
    const scale = f32(Math.sqrt(2.0 / 64.0))
    const d = new Float32Array(16 * 64)
    for (let i = 0; i < 16; i++) {
      for (let j = 0; j < 64; j++) {
        d[i * 64 + j] = f32(scale * f32(Math.cos((Math.PI / 2.0 / 64.0) * (i + 1) * (2 * j + 1))))
      }
    }
    return d
  })()

  // Torben median (no sort, matches the C++ reference exactly)
  function torben (m) {
    const n = m.length
    let lo = m[0]
    let hi = m[0]
    for (let i = 1; i < n; i++) {
      if (m[i] < lo) lo = m[i]
      if (m[i] > hi) hi = m[i]
    }
    const halfN = Math.floor((n + 1) / 2)
    let less = 0
    let greater = 0
    let guess = 0
    let maxlt = 0
    let mingt = 0
    for (;;) {
      guess = f32(f32(lo + hi) / 2)
      less = 0
      greater = 0
      maxlt = lo
      mingt = hi
      for (let i = 0; i < n; i++) {
        if (m[i] < guess) {
          less++
          if (m[i] > maxlt) maxlt = m[i]
        } else if (m[i] > guess) {
          greater++
          if (m[i] < mingt) mingt = m[i]
        }
      }
      if (less <= halfN && greater <= halfN) break
      if (less > greater) hi = maxlt
      else lo = mingt
    }
    const equal = n - less - greater
    if (less >= halfN) return maxlt
    if (less + equal >= halfN) return guess
    return mingt
  }

  // luma: rows*cols values 0..255 (plain Array or typed array, row-major).
  // Returns Uint8Array(32) — bit k (MSB byte 0) = DCT coeff (k/16, k%16) > median.
  function fromLuma (luma, width, height) {
    const rows = height
    const cols = width
    let b64
    if (rows === 64 && cols === 64) {
      b64 = Float32Array.from(luma)
    } else {
      const buf = Float32Array.from(luma)
      const wrow = Math.floor((cols + 127) / 128)
      const wcol = Math.floor((rows + 127) / 128)
      jarosz(buf, rows, cols, wrow, wcol)
      b64 = decimate(buf, rows, cols)
    }

    // 16x64 * 64x64 * 64x16 DCT. Sequential accumulation with the product
    // kept in double and one rounding per step (FMA semantics) reproduces
    // numpy's float32 sgemm bit-exactly on the reference vectors.
    const tmp = new Float32Array(16 * 64) // DCT @ b64
    for (let i = 0; i < 16; i++) {
      for (let j = 0; j < 64; j++) {
        let s = 0
        for (let k = 0; k < 64; k++) s = f32(s + DCT[i * 64 + k] * b64[k * 64 + j])
        tmp[i * 64 + j] = s
      }
    }
    const B = new Float32Array(16 * 16) // tmp @ DCT^T
    for (let i = 0; i < 16; i++) {
      for (let j = 0; j < 16; j++) {
        let s = 0
        for (let k = 0; k < 64; k++) s = f32(s + tmp[i * 64 + k] * DCT[j * 64 + k])
        B[i * 16 + j] = s
      }
    }

    const median = torben(B)
    const hash = new Uint8Array(32)
    for (let k = 0; k < 256; k++) {
      if (B[k] > median) hash[k >> 3] |= 1 << (7 - (k & 7))
    }
    return hash
  }

  // data: RGBA Uint8ClampedArray (canvas ImageData.data), REC601 luma
  function fromImageData (data, width, height) {
    const luma = new Float32Array(width * height)
    for (let p = 0; p < luma.length; p++) {
      const o = p * 4
      luma[p] = f32(f32(REC601[0] * data[o]) + f32(REC601[1] * data[o + 1]) + f32(REC601[2] * data[o + 2]))
    }
    return fromLuma(luma, width, height)
  }

  // Hamming distance over the first 104 bits (13 bytes) — the pHash slice
  function hamming104 (hashBytes, otherBytes) {
    let d = 0
    for (let i = 0; i < 13; i++) {
      let x = hashBytes[i] ^ otherBytes[i]
      while (x) {
        d += x & 1
        x >>= 1
      }
    }
    return d
  }

  return { fromLuma, fromImageData, hamming104 }
})()
